src/lib/config.rs

```diff
@@ -46,19 +53,72 @@ pub struct AppConfig {
     pub pub_cert: String,
     pub priv_cert: String,
 }
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct AppFileConfig {
+    pub log_level: LogLevel,
+    pub host: String,
+    pub port: u16,
+    pub pub_cert: String,
+    pub priv_cert: String,
+}
+
 pub fn load_config(service_name: String) -> Result<Config> {
     let args: CliArgument = CliArgument::parse();
 
     println!("Working directory {}", current_dir()?.display());
     println!("Loading config from {}", args.config.display());
+
     let config = std::fs::read_to_string(&args.config)?;
-    let mut config: Config = serde_json::from_str(&config)?;
-    config.app.log_level = args.log_level;
-    config.app.port = args.port;
-    config.app.host = args.host;
-    config.app.name = service_name;
-    config.app.pub_cert = args.pub_cert;
-    config.app.priv_cert = args.priv_cert;
-    println!("App config {:#?}", config.app);
-    Ok(config)
+    let root: serde_json::Value = serde_json::from_str(&config)?;
```

Comment:
I believe this simplification wouldn't work as it is since I removed the "name" field from `/etc/config.json`, but I could have misunderstood what you meant.

I could also put the "name" field back! 
I had removed it because it became redundant when I added separate configurations for each app, since the app name is the key of the entire app config object:
![current_app_config_example](https://user-images.githubusercontent.com/29493965/198750172-d5fe9814-3090-4a29-8532-dbc855f94c6b.png)

Adding the "name" field back to each app config could simplify the logic (AppFileConfig struct would no longer be needed), but the config would look like this:
![redundant_app_config_example](https://user-images.githubusercontent.com/29493965/198750182-e7655196-581e-42f5-ba4c-2b4bf1030d13.png)

Reply:
name field for each app doesn't matter. Previous the names are passed from main function. Actually I've implemented a better config in production code, and I'll post it here

Reply:
```rust

use crate::database::DatabaseConfig;
use crate::log::LogLevel;
use clap::Parser;
use eyre::*;
use serde::de::DeserializeOwned;
use serde::*;
use serde_json::Value;
use std::collections::HashMap;
use std::env::current_dir;
use std::fmt::Debug;
use std::path::PathBuf;

#[derive(Parser)]
#[clap(author, version, about, long_about = None)]
struct CliArgument {
    /// The path to config file
    #[clap(
        short,
        long,
        value_parser,
        value_name = "FILE",
        default_value = "etc/config.json",
        env = "CONFIG"
    )]
    config: PathBuf,
    #[clap(long, env = "LOG_LEVEL")]
    log_level: Option<LogLevel>,
    #[clap(long, env = "HOST")]
    /// The host to listen on
    host: Option<String>,
    #[clap(long, env = "PORT")]
    /// The port to listen on
    port: Option<u16>,
    #[clap(long, env = "DEBUG")]
    debug: bool,
    #[clap(long, env = "PUB_CERTS", value_delimiter = ',')]
    pub_certs: Option<Vec<String>>,
    #[clap(long, env = "PRIV_CERT")]
    priv_cert: Option<String>,
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Config<App> {
    pub app_db: DatabaseConfig,
    pub auth_db: DatabaseConfig,
    #[serde(skip)]
    pub app: AppConfig<App>,
    #[serde(flatten)]
    pub apps: HashMap<String, AppConfig<Value>>,
}
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct AppConfig<App> {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub log_level: LogLevel,
    #[serde(default)]
    pub host: String,
    #[serde(default)]
    pub port: u16,
    #[serde(default)]
    pub pub_certs: Option<Vec<String>>,
    #[serde(default)]
    pub priv_cert: Option<String>,
    #[serde(default)]
    pub debug: bool,
    #[serde(skip)]
    pub header_only: bool,
    #[serde(flatten)]
    pub extra: App,
}
pub fn load_config<App: DeserializeOwned + Debug + Default>(
    service_name: String,
) -> Result<Config<App>> {
    let args: CliArgument = CliArgument::parse();

    println!("Working directory {}", current_dir()?.display());
    println!("Loading config from {}", args.config.display());
    let config = std::fs::read_to_string(&args.config)?;
    let mut config: Config<App> = serde_json::from_str(&config)?;
    if let Some(a) = config.apps.get(&service_name) {
        config.app = serde_json::from_value(serde_json::to_value(a)?)?;
    }

    // merge configs
    if let Some(log_level) = args.log_level {
        config.app.log_level = log_level;
    }

    if let Some(host) = args.host {
        config.app.host = host;
    }
    if let Some(port) = args.port {
        config.app.port = port;
    }
    config.app.name = service_name;
    if let Some(pub_certs) = args.pub_certs {
        config.app.pub_certs = Some(pub_certs);
    }
    if let Some(priv_cert) = args.priv_cert {
        config.app.priv_cert = Some(priv_cert);
    }
    if args.debug {
        config.app.debug = true;
    }
    println!("App config {:#?}", config.app);
    Ok(config)
}

```

Reply:
thanks man, already fixed and pushed to local-storage branch

I added the better version of the config.rs following this file, I removed the generics, and had to change a couple of things in `ws/server`